--- wmmp3-0.11/main.c.pius	Fri Dec 10 18:10:13 1999
+++ wmmp3-0.11/main.c	Fri Dec 10 18:15:20 1999
@@ -114,7 +114,7 @@
 	errno = 0;
 
 	/* set defualts in case anything fails */
-	set_mpg123("/usr/local/bin/mpg123");
+	set_mpg123("/usr/bin/mpg123");
 	set_mp3ext(".mp3");
 	set_playlistext(".m3u");
 
--- wmmp3-0.11/sample.wmmp3.pius	Thu Dec  9 05:25:25 1999
+++ wmmp3-0.11/sample.wmmp3	Fri Dec 10 18:30:45 1999
@@ -1,5 +1,5 @@
 # full path to mpg123 executable
-mpg123 = /usr/local/bin/mpg123
+mpg123 = /usr/bin/mpg123
 
 #
 # mp3 directories, 1 per line
